devDocs/startupShutdownInstall.md

```diff
@@ -0,0 +1,96 @@
+    1. NVDA+q
+        - test: `startupShutdownNVDA.Quits from keyboard, Restarts`
+    1. An unbound command to restart
+    1. A prompt to restart from changing certain settings
+    1. Via the NVDA menu -> Exit
+        - test: `startupShutdownNVDA.Quits from menu`
+1. A process sending `WM_QUIT`, eg a new NVDA process starting (uses `WM_QUIT`)
+1. A handled crash (directly causes a new process to start, terminates unsafely)
+    - test: `startupShutdownNVDA.Restarts on crash, Restarts on braille crash`
+1. An unhandled crash (terminates unsafely)
+1. An external command which kills the process (terminates unsafely) 
+1. Windows shutting down (terminates unsafely)
+    - On `wx.EVT_END_SESSION`, we save the config and play the exit sound. (Note: should this do more? NVDA may terminate unsafely directly after this)
+
+## Safe exit hooks/triggers
+
+### In process: `triggerNVDAExit`
+* A function in the core module
+* Only executes the code once, uses a lock and flag to ensure this
+* Uses a queue on the main thread to queue a safe shutdown
+* Once the queued shutdown starts:
+    1. the updateCheck is terminated
+    1. watchdog is terminated
+    1. addons and the brailleViewer are terminated, so we can close all windows safely
+    1. All NVDA windows are closed
+    1. Now that windows are closed, a new NVDA instance is started if requested
+
+
+### Out of process: `WM_QUIT`
+* a windows message which can be sent across processes which will force the main loop to exit and close most NVDA windows.
+* We subsequently run `triggerNVDAExit` to exit other required processes, and pump the queue to execute the shutdown.
+
+## Technical startup and exit process
+
+All the startup pathways execute nvda.pyw.
+
+1. *Start nvda.pyw* 
+1. MonkeyPatches are applied
+1. The virtual environment is checked
+1. Gettext is installed
+1. Starting arguments are parsed
+1. We check for a running instance of NVDA by looking for a Message Window
+    - terminate it if possible using `WM_QUIT` or forcing the process to exit. If this fails, abandon startup
+1. We try to acqure the NVDA mutex
+    - abandon startup if we can't
+1. Initialise logging
+1. Set the process to allow `WM_QUIT` messages
+1. Start running `core.main()`
+    1. SetProcessDPIAware
+    1. Initialize config
+    1. Play statup sound
+    1. Set language
+    1. Initialize NVDA modules and wx GUI
+    1. Start the Message Window (used for `WM_QUIT`)
+    1. Start the core pump
+    1. Start watchdog
+    1. Notify postNvdaStartup
+    1. Start the wx app main loop
+    --- 
+    1. *The main loop has exited via `triggerNVDAExit` or `WM_QUIT`*
+    1. If `WM_QUIT` was used to exit the main loop, perform `triggerNVDAExit`
+    1. Terminate gui
+    1. Save config
+    1. Terminate NVDA modules
+    1. Play exit sound
+    1. Destroy the message window
+    1. Return to nvda.pyw
+1. Notify ease of access of the shutdown
+1. Release the Mutex
+1. Exit nvda.pyw with `sys.exit`

```

Comment:
I'd avoid this level of detail. It will be hard to keep it in sync with the code. Instead I'd focus on the key points:
- NVDA protects itself from having multiple instances running:
  - By detecting the presence of a Window Class, in this case WM_QUIT is used to signal the old process.
  - By waiting to acquire an OS mutex.

Reply:
Which aspects do you think are unnecessary? I'm going to suggest grouping some but I do think these are all fairly core parts of the shutdown/start process

Reply:
The code serves as a better reference. Keeping this accurate (in sync) with code will be difficult. This isn't necessarily highlighting information that isn't already in the code, low signal to noise ratio.

Instead I'd focus on key aspects of the code that may not be immediately obvious. If you can draw attention to facts about the way the code works, it becomes easier to reason about and debug when there are issues. For instance, after the mutex is acquired there can not be another copy of NVDA running in that "critical section" (which extends until the final moments of shutting down).